File: src/types/magazzino.ts
// Tipi per il modulo Magazzino

// === CATEGORIE PRODOTTI ===

export interface CategoriaProdotto {
  id: string;
  codice: string;
  nome: string;
  tipo: 'consumo' | 'rivendita' | 'entrambi' | null;
  created_at: string;
}

export interface CreateCategoriaProdottoInput {
  nome: string;
  tipo: 'consumo' | 'rivendita' | 'entrambi';
}

export interface UpdateCategoriaProdottoInput {
  nome?: string;
  tipo?: 'consumo' | 'rivendita' | 'entrambi';
}

// === PRODOTTI ===

export interface Prodotto {
  id: string;
  codice: string;
  barcode: string | null;
  categoria_id: string | null;
  nome: string;
  descrizione: string | null;
  marca: string | null;
  linea: string | null;
  unita_misura: string;
  capacita: number | null;
  giacenza: number;
  scorta_minima: number;
  scorta_riordino: number;
  prezzo_acquisto: number | null;
  prezzo_vendita: number | null;
  uso: 'interno' | 'vendita' | 'entrambi' | null;
  attivo: boolean;
  data_scadenza: string | null;
  note: string | null;
  created_at: string;
  updated_at: string;
  // Campi join
  categoria_nome?: string | null;
  categoria_tipo?: string | null;
}

export interface CreateProdottoInput {
  nome: string;
  categoria_id?: string;
  barcode?: string;
  descrizione?: string;
  marca?: string;
  linea?: string;
  unita_misura?: string;
  capacita?: number;
  giacenza?: number;
  scorta_minima?: number;
  scorta_riordino?: number;
  prezzo_acquisto?: number;
  prezzo_vendita?: number;
  uso?: 'interno' | 'vendita' | 'entrambi';
  data_scadenza?: string;
  note?: string;
}

export interface UpdateProdottoInput {
  nome?: string;
  categoria_id?: string;
  barcode?: string;
  descrizione?: string;
  marca?: string;
  linea?: string;
  unita_misura?: string;
  capacita?: number;
  scorta_minima?: number;
  scorta_riordino?: number;
  prezzo_acquisto?: number;
  prezzo_vendita?: number;
  uso?: 'interno' | 'vendita' | 'entrambi';
  data_scadenza?: string;
  note?: string;
  attivo?: boolean;
}

// === MOVIMENTI MAGAZZINO ===

export type TipoMovimento = 'carico' | 'scarico_uso' | 'scarico_vendita' | 'reso' | 'inventario' | 'scarto';

export interface MovimentoMagazzino {
  id: string;
  prodotto_id: string;
  tipo: TipoMovimento;
  quantita: number;
  giacenza_risultante: number;
  appuntamento_id: string | null;
  operatrice_id: string | null;
  cliente_id: string | null;
  fornitore: string | null;
  documento_riferimento: string | null;
  prezzo_unitario: number | null;
  lotto: string | null;
  data_scadenza: string | null;
  note: string | null;
  created_at: string;
  // Campi join
  prodotto_nome?: string | null;
  prodotto_codice?: string | null;
  operatrice_nome?: string | null;
  cliente_nome?: string | null;
}

export interface CreateCaricoInput {
  prodotto_id: string;
  quantita: number;
  fornitore?: string;
  documento_riferimento?: string;
  prezzo_unitario?: number;
  lotto?: string;
  data_scadenza?: string;
  note?: string;
}

export interface CreateScaricoInput {
  prodotto_id: string;
  quantita: number;
  tipo: 'scarico_uso' | 'scarico_vendita' | 'scarto';
  operatrice_id?: string;
  cliente_id?: string;
  appuntamento_id?: string;
  note?: string;
}

export interface CreateResoInput {
  prodotto_id: string;
  quantita: number;
  operatrice_id?: string;
  cliente_id?: string;
  appuntamento_id?: string;
  note?: string;
}

export interface CreateInventarioInput {
  prodotto_id: string;
  nuova_giacenza: number;
  note?: string;
}

// === FILTRI ===

export interface FiltriMovimenti {
  prodotto_id?: string;
  tipo?: TipoMovimento;
  data_da?: string;
  data_a?: string;
  operatrice_id?: string;
  cliente_id?: string;
  fornitore?: string;
}

// === ALERT ===

export interface AlertProdotto {
  id: string;
  codice: string;
  nome: string;
  tipo_alert: 'scorta_minima' | 'scadenza_vicina' | 'scaduto';
  giacenza: number;
  scorta_minima: number;
  data_scadenza: string | null;
  giorni_alla_scadenza: number | null;
}

export interface AlertCount {
  sotto_scorta: number;
  in_scadenza: number;
  scaduti: number;
}

// === REPORT ===

export interface ReportConsumiResult {
  prodotto_id: string;
  prodotto_nome: string;
  prodotto_codice: string;
  categoria_nome: string | null;
  quantita_totale: number;
  valore_totale: number | null;
  numero_movimenti: number;
}

export interface ValoreMagazzino {
  valore_acquisto: number | null;
  valore_vendita: number | null;
  totale_prodotti: number;
  totale_pezzi: number;
}

// === HELPER TYPES ===

export const TIPI_MOVIMENTO_LABELS: Record<TipoMovimento, string> = {
  carico: 'Carico',
  scarico_uso: 'Scarico Uso',
  scarico_vendita: 'Vendita',
  reso: 'Reso',
  inventario: 'Inventario',
  scarto: 'Scarto',
};

export const UNITA_MISURA_OPTIONS = [
  { value: 'pz', label: 'Pezzi' },
  { value: 'ml', label: 'Millilitri' },
  { value: 'l', label: 'Litri' },
  { value: 'g', label: 'Grammi' },
  { value: 'kg', label: 'Kilogrammi' },
  { value: 'conf', label: 'Confezioni' },
];

export const USO_OPTIONS = [
  { value: 'interno', label: 'Uso Interno' },
  { value: 'vendita', label: 'Rivendita' },
  { value: 'entrambi', label: 'Entrambi' },
];

export const TIPO_CATEGORIA_OPTIONS = [
  { value: 'consumo', label: 'Consumo' },
  { value: 'rivendita', label: 'Rivendita' },
  { value: 'entrambi', label: 'Entrambi' },
];

// === INVENTARIO ===

export type StatoInventario = 'in_corso' | 'confermato' | 'annullato';

export interface Inventario {
  id: string;
  codice: string;
  descrizione: string | null;
  data_inizio: string;
  data_chiusura: string | null;
  stato: StatoInventario;
  note: string | null;
  created_at: string;
  updated_at: string;
}

export interface RigaInventario {
  id: string;
  inventario_id: string;
  prodotto_id: string;
  giacenza_teorica: number;
  quantita_contata: number;
  differenza: number;
  lotto: string | null;
  data_scadenza: string | null;
  note: string | null;
  created_at: string;
  updated_at: string;
}

export interface RigaInventarioWithProdotto extends RigaInventario {
  prodotto_codice: string | null;
  prodotto_nome: string | null;
  prodotto_barcode: string | null;
  prodotto_marca: string | null;
  prodotto_unita_misura: string | null;
}

export interface CreateSessioneInventarioInput {
  descrizione?: string;
  note?: string;
}

export interface CreateRigaInventarioInput {
  inventario_id: string;
  prodotto_id: string;
  quantita_contata: number;
  lotto?: string;
  data_scadenza?: string;
  note?: string;
}

export interface UpdateRigaInventarioInput {
  quantita_contata?: number;
  lotto?: string;
  data_scadenza?: string;
  note?: string;
}

export interface InventarioRiepilogo {
  totale_righe: number;
  prodotti_contati: number;
  differenze_positive: number;
  differenze_negative: number;
  valore_differenza: number | null;
}

export interface ProdottoPerInventario {
  id: string;
  codice: string;
  barcode: string | null;
  nome: string;
  marca: string | null;
  unita_misura: string;
  giacenza: number;
  gia_inserito: boolean;
}

export const STATO_INVENTARIO_LABELS: Record<StatoInventario, string> = {
  in_corso: 'In Corso',
  confermato: 'Confermato',
  annullato: 'Annullato',
};

export const STATO_INVENTARIO_COLORS: Record<StatoInventario, string> = {
  in_corso: 'bg-yellow-50 text-yellow-700',
  confermato: 'bg-green-50 text-green-700',
  annullato: 'bg-gray-100 text-gray-600',
};

File: src-tauri/src/models/magazzino.rs
// Modelli per il modulo Magazzino
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use chrono::NaiveDateTime;

// === CATEGORIE PRODOTTI ===

#[derive(Debug, Serialize, Deserialize, FromRow, Clone)]
pub struct CategoriaProdotto {
    pub id: String,
    pub codice: String,
    pub nome: String,
    pub tipo: Option<String>, // 'consumo' | 'rivendita' | 'entrambi'
    pub created_at: NaiveDateTime,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateCategoriaProdottoInput {
    pub nome: String,
    pub tipo: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateCategoriaProdottoInput {
    pub nome: Option<String>,
    pub tipo: Option<String>,
}

// === PRODOTTI ===

#[derive(Debug, Serialize, Deserialize, FromRow, Clone)]
pub struct ProdottoWithCategoria {
    pub id: String,
    pub codice: String,
    pub barcode: Option<String>,
    pub categoria_id: Option<String>,
    pub nome: String,
    pub descrizione: Option<String>,
    pub marca: Option<String>,
    pub linea: Option<String>,
    pub unita_misura: String,
    pub capacita: Option<f64>,
    pub giacenza: f64,
    pub scorta_minima: f64,
    pub scorta_riordino: f64,
    pub prezzo_acquisto: Option<f64>,
    pub prezzo_vendita: Option<f64>,
    pub uso: Option<String>,
    pub attivo: bool,
    pub data_scadenza: Option<String>,  // String per gestire valori vuoti nel DB
    pub note: Option<String>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
    // Join fields
    pub categoria_nome: Option<String>,
    pub categoria_tipo: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateProdottoInput {
    pub nome: String,
    pub categoria_id: Option<String>,
    pub barcode: Option<String>,
    pub descrizione: Option<String>,
    pub marca: Option<String>,
    pub linea: Option<String>,
    pub unita_misura: Option<String>,
    pub capacita: Option<f64>,
    pub giacenza: Option<f64>,
    pub scorta_minima: Option<f64>,
    pub scorta_riordino: Option<f64>,
    pub prezzo_acquisto: Option<f64>,
    pub prezzo_vendita: Option<f64>,
    pub uso: Option<String>,
    pub data_scadenza: Option<String>,
    pub note: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateProdottoInput {
    pub nome: Option<String>,
    pub categoria_id: Option<String>,
    pub barcode: Option<String>,
    pub descrizione: Option<String>,
    pub marca: Option<String>,
    pub linea: Option<String>,
    pub unita_misura: Option<String>,
    pub capacita: Option<f64>,
    pub scorta_minima: Option<f64>,
    pub scorta_riordino: Option<f64>,
    pub prezzo_acquisto: Option<f64>,
    pub prezzo_vendita: Option<f64>,
    pub uso: Option<String>,
    pub data_scadenza: Option<String>,
    pub note: Option<String>,
    pub attivo: Option<bool>,
}

// === MOVIMENTI MAGAZZINO ===

#[derive(Debug, Serialize, Deserialize, FromRow, Clone)]
pub struct MovimentoMagazzino {
    pub id: String,
    pub prodotto_id: String,
    pub tipo: String, // 'carico' | 'scarico_uso' | 'scarico_vendita' | 'reso' | 'inventario' | 'scarto'
    pub quantita: f64,
    pub giacenza_risultante: f64,
    pub appuntamento_id: Option<String>,
    pub operatrice_id: Option<String>,
    pub cliente_id: Option<String>,
    pub fornitore: Option<String>,
    pub documento_riferimento: Option<String>,
    pub prezzo_unitario: Option<f64>,
    pub lotto: Option<String>,
    pub data_scadenza: Option<String>,
    pub note: Option<String>,
    pub created_at: NaiveDateTime,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Clone)]
pub struct MovimentoWithDetails {
    pub id: String,
    pub prodotto_id: String,
    pub tipo: String,
    pub quantita: f64,
    pub giacenza_risultante: f64,
    pub appuntamento_id: Option<String>,
    pub operatrice_id: Option<String>,
    pub cliente_id: Option<String>,
    pub fornitore: Option<String>,
    pub documento_riferimento: Option<String>,
    pub prezzo_unitario: Option<f64>,
    pub lotto: Option<String>,
    pub data_scadenza: Option<String>,
    pub note: Option<String>,
    pub created_at: NaiveDateTime,
    // Join fields
    pub prodotto_nome: Option<String>,
    pub prodotto_codice: Option<String>,
    pub operatrice_nome: Option<String>,
    pub cliente_nome: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateCaricoInput {
    pub prodotto_id: String,
    pub quantita: f64,
    pub fornitore: Option<String>,
    pub documento_riferimento: Option<String>,
    pub prezzo_unitario: Option<f64>,
    pub lotto: Option<String>,
    pub data_scadenza: Option<String>,
    pub note: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateScaricoInput {
    pub prodotto_id: String,
    pub quantita: f64,
    pub tipo: String, // 'scarico_uso' | 'scarico_vendita' | 'scarto'
    pub operatrice_id: Option<String>,
    pub cliente_id: Option<String>,
    pub appuntamento_id: Option<String>,
    pub note: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateInventarioInput {
    pub prodotto_id: String,
    pub nuova_giacenza: f64,
    pub note: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateResoInput {
    pub prodotto_id: String,
    pub quantita: f64,
    pub operatrice_id: Option<String>,
    pub cliente_id: Option<String>,
    pub appuntamento_id: Option<String>,
    pub note: Option<String>,
}

// === FILTRI ===

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct FiltriMovimenti {
    pub prodotto_id: Option<String>,
    pub tipo: Option<String>,
    pub data_da: Option<String>,
    pub data_a: Option<String>,
    pub operatrice_id: Option<String>,
    pub cliente_id: Option<String>,
    pub fornitore: Option<String>,
}

// === ALERT E REPORT ===

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct AlertProdotto {
    pub id: String,
    pub codice: String,
    pub nome: String,
    pub tipo_alert: String, // 'scorta_minima' | 'scadenza_vicina' | 'scaduto'
    pub giacenza: f64,
    pub scorta_minima: f64,
    pub data_scadenza: Option<String>,
    pub giorni_alla_scadenza: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AlertCount {
    pub sotto_scorta: i64,
    pub in_scadenza: i64,
    pub scaduti: i64,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct ReportConsumiResult {
    pub prodotto_id: String,
    pub prodotto_nome: String,
    pub prodotto_codice: String,
    pub categoria_nome: Option<String>,
    pub quantita_totale: f64,
    pub valore_totale: Option<f64>,
    pub numero_movimenti: i64,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct ValoreMagazzino {
    pub valore_acquisto: Option<f64>,
    pub valore_vendita: Option<f64>,
    pub totale_prodotti: i64,
    pub totale_pezzi: f64,
}

// === INVENTARIO ===

#[derive(Debug, Serialize, Deserialize, FromRow, Clone)]
pub struct Inventario {
    pub id: String,
    pub codice: String,
    pub descrizione: Option<String>,
    pub data_inizio: NaiveDateTime,
    pub data_chiusura: Option<NaiveDateTime>,
    pub stato: String,  // 'in_corso' | 'confermato' | 'annullato'
    pub note: Option<String>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Clone)]
pub struct RigaInventario {
    pub id: String,
    pub inventario_id: String,
    pub prodotto_id: String,
    pub giacenza_teorica: f64,
    pub quantita_contata: f64,
    pub differenza: f64,
    pub lotto: Option<String>,
    pub data_scadenza: Option<String>,
    pub note: Option<String>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Clone)]
pub struct RigaInventarioWithProdotto {
    pub id: String,
    pub inventario_id: String,
    pub prodotto_id: String,
    pub giacenza_teorica: f64,
    pub quantita_contata: f64,
    pub differenza: f64,
    pub lotto: Option<String>,
    pub data_scadenza: Option<String>,
    pub note: Option<String>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
    // Join fields
    pub prodotto_codice: Option<String>,
    pub prodotto_nome: Option<String>,
    pub prodotto_barcode: Option<String>,
    pub prodotto_marca: Option<String>,
    pub prodotto_unita_misura: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateSessioneInventarioInput {
    pub descrizione: Option<String>,
    pub note: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateRigaInventarioInput {
    pub inventario_id: String,
    pub prodotto_id: String,
    pub quantita_contata: f64,
    pub lotto: Option<String>,
    pub data_scadenza: Option<String>,
    pub note: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateRigaInventarioInput {
    pub quantita_contata: Option<f64>,
    pub lotto: Option<String>,
    pub data_scadenza: Option<String>,
    pub note: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct InventarioRiepilogo {
    pub id: String,
    pub codice: String,
    pub descrizione: Option<String>,
    pub data_inizio: NaiveDateTime,
    pub stato: String,
    pub totale_righe: i64,
    pub totale_differenze_positive: f64,
    pub totale_differenze_negative: f64,
}
